Improve error log in k8s discovery

vinicius-colutti opened the pull request:
Add ServiceAccount in KubernetesClient to show the specific serviceAccount and namespace, for Forbidden error

<!--
Contributing to Hazelcast and looking for a challenge? Why don't you check out our open positions?

https://hazelcast.pinpointhq.com/
-->

I am creating this PR, to solve this issue, all details are written in the issue

Fixes https://github.com/hazelcast/hazelcast/issues/24706

Backport of: INSERT_LINK_TO_THE_ORIGINAL_PR_HERE

EE PR: INSERT_LINK_TO_THE_EE_PR_HERE

Breaking changes (list specific methods/types/messages):
* API
* client protocol format
* serialized form
* snapshot format

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org:443/pull/badge/not_signed)](https://cla.hazelcast.org:443/hazelcast/hazelcast?pullRequest=25568) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.hazelcast.org:443/hazelcast/hazelcast?pullRequest=25568) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org:443/check/hazelcast/hazelcast?pullRequest=25568) it.</sub>

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

cagric0 commented:
Hi @vinicius-colutti, Thanks for your effort and the PR.

As you mentioned, sometimes the logs are not enough to understand the issue. For just printing serviceAccount, we are not planning to take it in the client config. The user is already knew which serviceAccount is used in which namespace(created by himself for the client). But as you suggest, printing additional logs would be helpful to understand the issue(missing permission in serviceAccount). 

Here is the updated version I am thinking:
```
LOGGER.warning("Kubernetes API access is forbidden! Starting standalone. To use Hazelcast Kubernetes discovery,"
 + " configure the required RBAC. For 'default' service account in 'default' namespace execute: "
 + "`kubectl apply -f https://raw.githubusercontent.com/hazelcast/hazelcast/master/kubernetes-rbac.yaml`" 
 + "If you want to use different service account and different namespace, you can update the mentioned rbac.yaml file accordingly and use it");

LOGGER.warning("Error Kubernetes API Cause: ", e.getCause());
LOGGER.warning("Error Kubernetes API Cause message: "+ e.getCause().getMessage());
```

vinicius-colutti commented:
Hi @cagric0 Thanks for your review, I added the serviceAccount in the clientConfig, because I have a specific situation and sometimes users don't recognize their serviceAccount or namespace, but I know and understand now that this situation is not common.

So I agree with your suggestion and changed this PR with these changes, thank you!!

kwart commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/21170/Hazelcast-pr-builder-21170.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/21170/Hazelcast-pr-builder-21170.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesApiEndpointResolver.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:620:1: File contains tab characters (this is the first instance). [FileTabCharacter]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:621: Line has trailing spaces. [RegexpSingleline]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:621: Line is longer than 130 characters (found 138). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:622: Line is longer than 130 characters (found 175). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:624:86: '+' is not preceded with whitespace. [WhitespaceAround]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:625: Line has trailing spaces. [RegexpSingleline]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesConfig.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesProperties.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/kubernetes/KubernetesClientTest.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/test/java/com/hazelcast/kubernetes/StsMonitorTest.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.3.0:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 12 errors reported by Checkstyle 9.3 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
[ERROR] 
--------------------------
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
--------------------------
[ERROR] 
--------------------------
[ERROR] For more information about the errors and possible solutions, please read the following articles:
--------------------------
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
--------------------------
[ERROR] 
--------------------------
[ERROR] After correcting the problems, you can resume the build with the command
--------------------------
[ERROR]   mvn <args> -rf :hazelcast
--------------------------
</pre>
</details>

kwart commented on hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesApiEndpointResolver.java:
```diff
@@ -190,4 +190,4 @@ void start() {
         client.destroy();
     }
-}
+}
```
Why do you change the file end here?
This should be reverted - including other places with the same change.

kwart commented on hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:
```diff
@@ -617,8 +617,12 @@ private List<Endpoint> handleKnownException(RestClientException e) {
                 LOGGER.warning("Kubernetes API access is forbidden! Starting standalone. To use Hazelcast Kubernetes discovery,"
-                        + " configure the required RBAC. For 'default' service account in 'default' namespace execute: "
-                        + "`kubectl apply -f https://raw.githubusercontent.com/hazelcast/hazelcast/master/kubernetes-rbac.yaml`");
+				 + " configure the required RBAC. For 'default' service account in 'default' namespace execute: "
+				 + "`kubectl apply -f https://raw.githubusercontent.com/hazelcast/hazelcast/master/kubernetes-rbac.yaml`" 
+				 + "If you want to use different service account and different namespace, you can update the mentioned rbac.yaml file accordingly and use it");
+				LOGGER.warning("Error Kubernetes API Cause: ", e.getCause());
+				LOGGER.warning("Error Kubernetes API Cause message: "+ e.getCause().getMessage());
```
The formatting doesn't seem to be correct here. Moreover, it doesn't make sense to split the log into 3 messages. Could you use just the first one, but add the `e` as last parameter to the `warning(..., e);` method?

kwart reviewed (changes requested):
Thank you for adding more details with solution description to the log message. It's a nice improvement. I suggest to use just the single log message and revert unrelated formatting changes.

vinicius-colutti reviewed (commented):
Hi @kwart  Thanks for your review.

Can you check again?

kwart commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/21668/Hazelcast-pr-builder-21668.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/21668/Hazelcast-pr-builder-21668.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:621:1: File contains tab characters (this is the first instance). [FileTabCharacter]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java:622: Line is longer than 130 characters (found 138). [LineLength]
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.3.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 9.3 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
[ERROR] 
--------------------------
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
--------------------------
[ERROR] 
--------------------------
[ERROR] For more information about the errors and possible solutions, please read the following articles:
--------------------------
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
--------------------------
[ERROR] 
--------------------------
[ERROR] After correcting the problems, you can resume the build with the command
--------------------------
[ERROR]   mvn <args> -rf :hazelcast
--------------------------
</pre>
</details>

vinicius-colutti commented:
run-lab-run

kwart commented:
run-lab-run

kwart commented:
Merged. Thanks for the fix, @vinicius-colutti !
